File: remote/src/test/java/lt/markmerkk/JiraLogFilterFiltererTest.java
package lt.markmerkk;

import java.util.Date;
import java.util.List;
import net.rcarz.jiraclient.User;
import net.rcarz.jiraclient.WorkLog;
import org.joda.time.DateTime;
import org.junit.Before;
import org.junit.Test;
import rx.Observable;
import rx.observers.TestSubscriber;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;

/**
 * Created by mariusmerkevicius on 1/30/16.
 */
public class JiraLogFilterFiltererTest {
  private User user;
  private WorkLog worklog;
  private TestSubscriber<WorkLog> testSubscriber;

  @Before
  public void setUp() {
    testSubscriber = new TestSubscriber<>();
    user = mock(User.class);
    worklog = mock(WorkLog.class);
    DateTime worklogTime = new DateTime(1000);
    doReturn(new Date(worklogTime.getMillis())).when(worklog).getStarted();
    doReturn(user).when(worklog).getAuthor();
    doReturn("asdf").when(user).getName();
  }

  @Test
  public void filterLog_valid_shouldReturnWorklog() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer("asdf", new DateTime(500), new DateTime(2000));

    // Act
    WorkLog out = filterer.filterLog(worklog);

    // Assert
    assertThat(out).isNotNull();
  }

  @Test
  public void filterLog_endEqualCreate_shouldReturnWorklog() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer("asdf", new DateTime(500), new DateTime(1000));

    // Act
    WorkLog out = filterer.filterLog(worklog);

    // Assert
    assertThat(out).isNotNull();
  }

  @Test
  public void filterLog_startEqualCreate_shouldReturnWorklog() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer("asdf", new DateTime(1000), new DateTime(2000));

    // Act
    WorkLog out = filterer.filterLog(worklog);

    // Assert
    assertThat(out).isNotNull();
  }

  @Test
  public void filterLog_endBeforeCreate_shouldReturnNull() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer("asdf", new DateTime(500), new DateTime(900));

    // Act
    WorkLog out = filterer.filterLog(worklog);

    // Assert
    assertThat(out).isNull();
  }

  @Test
  public void filterLog_startAfterCreate_shouldReturnNull() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer("asdf", new DateTime(1200), new DateTime(2000));

    // Act
    WorkLog out = filterer.filterLog(worklog);

    // Assert
    assertThat(out).isNull();
  }

  @Test
  public void filterLog_invalidUser_shouldReturnNull() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer("null", new DateTime(500), new DateTime(2000));

    // Act
    WorkLog out = filterer.filterLog(worklog);

    // Assert
    assertThat(out).isNull();
  }

  @Test
  public void filterLog_nullUser_shouldReturnNull() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer(null, new DateTime(500), new DateTime(2000));

    // Act
    WorkLog out = filterer.filterLog(worklog);

    // Assert
    assertThat(out).isNull();
  }

  @Test
  public void filterLog_nullStartTime_shouldReturnNull() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer("asdf", null, new DateTime(2000));

    // Act
    WorkLog out = filterer.filterLog(worklog);

    // Assert
    assertThat(out).isNull();
  }

  @Test
  public void filterLog_nullEndTime_shouldReturnNull() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer("asdf", new DateTime(500), null);

    // Act
    WorkLog out = filterer.filterLog(worklog);

    // Assert
    assertThat(out).isNull();
  }

  @Test
  public void filterLog_nullWorkLog_shouldReturnNull() throws Exception {
    // Arrange
    JiraLogFilterer filterer = new JiraLogFilterer("asdf", new DateTime(500), new DateTime(2000));

    // Act
    WorkLog out = filterer.filterLog(null);

    // Assert
    assertThat(out).isNull();
  }
}

File: remote/src/test/java/lt/markmerkk/JiraLogFilterFiltererTest.kt
package lt.markmerkk

import java.util.Date
import net.rcarz.jiraclient.User
import net.rcarz.jiraclient.WorkLog
import org.joda.time.DateTime
import org.junit.Before
import org.junit.Test
import rx.Observable
import rx.observers.TestSubscriber

import org.assertj.core.api.Assertions.assertThat
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.mock
import kotlin.test.assertNotNull
import kotlin.test.assertNull

/**
 * Created by mariusmerkevicius on 1/30/16.
 */
class JiraLogFilterFiltererTest {
    private var user: User? = null
    private var worklog: WorkLog? = null

    @Before
    fun setUp() {
        user = mock(User::class.java)
        worklog = mock(WorkLog::class.java)
        val worklogTime = DateTime(1000)
        doReturn(Date(worklogTime.millis)).`when`<WorkLog>(worklog).started
        doReturn(user).`when`<WorkLog>(worklog).author
        doReturn("asdf").`when`<User>(user).name
    }

    @Test
    fun filterLog_valid_shouldReturnWorklog() {
        // Arrange
        val filterer = JiraLogFilterer("asdf", DateTime(500), DateTime(2000))

        // Act
        val out = filterer.filterLog(worklog)

        // Assert
        assertNotNull(out)
    }

    @Test
    fun filterLog_endEqualCreate_shouldReturnWorklog() {
        // Arrange
        val filterer = JiraLogFilterer("asdf", DateTime(500), DateTime(1000))

        // Act
        val out = filterer.filterLog(worklog)

        // Assert
        assertNotNull(out)
    }

    @Test
    fun filterLog_startEqualCreate_shouldReturnWorklog() {
        // Arrange
        val filterer = JiraLogFilterer("asdf", DateTime(1000), DateTime(2000))

        // Act
        val out = filterer.filterLog(worklog)

        // Assert
        assertNotNull(out)
    }

    @Test
    fun filterLog_endBeforeCreate_shouldReturnNull() {
        // Arrange
        val filterer = JiraLogFilterer("asdf", DateTime(500), DateTime(900))

        // Act
        val out = filterer.filterLog(worklog)

        // Assert
        assertNull(out)
    }

    @Test
    fun filterLog_startAfterCreate_shouldReturnNull() {
        // Arrange
        val filterer = JiraLogFilterer("asdf", DateTime(1200), DateTime(2000))

        // Act
        val out = filterer.filterLog(worklog)

        // Assert
        assertNull(out)
    }

    @Test
    fun filterLog_invalidUser_shouldReturnNull() {
        // Arrange
        val filterer = JiraLogFilterer("null", DateTime(500), DateTime(2000))

        // Act
        val out = filterer.filterLog(worklog)

        // Assert
        assertNull(out)
    }

    @Test
    fun filterLog_nullUser_shouldReturnNull() {
        // Arrange
        val filterer = JiraLogFilterer(null, DateTime(500), DateTime(2000))

        // Act
        val out = filterer.filterLog(worklog)

        // Assert
        assertNull(out)
    }

    @Test
    fun filterLog_nullStartTime_shouldReturnNull() {
        // Arrange
        val filterer = JiraLogFilterer("asdf", null, DateTime(2000))

        // Act
        val out = filterer.filterLog(worklog)

        // Assert
        assertNull(out)
    }

    @Test
    fun filterLog_nullEndTime_shouldReturnNull() {
        // Arrange
        val filterer = JiraLogFilterer("asdf", DateTime(500), null)

        // Act
        val out = filterer.filterLog(worklog)

        // Assert
        assertNull(out)
    }

    @Test
    fun filterLog_nullWorkLog_shouldReturnNull() {
        // Arrange
        val filterer = JiraLogFilterer("asdf", DateTime(500), DateTime(2000))

        // Act
        val out = filterer.filterLog(null)

        // Assert
        assertNull(out)
    }
}